pep-0544.txt

```diff
@@ -0,0 +1,949 @@
+(non-protocol) subtypes of ``Proto``. For example::
+
+  class Proto(Protocol):
+      @abstractmethod
+      def meth(self) -> int:
+          ...
+  class Concrete:
+      def meth(self) -> int:
+          return 42
+
+  def fun(cls: Type[Proto]) -> int:
+      return cls().meth() # OK
+  fun(Proto)              # Error
+  fun(Concrete)           # OK
+
+The same rule applies to variables::
+
+  var: Type[Proto]
+  var = Proto    # Error
+  var = Concrete # OK
+  var().meth()   # OK
+
+Assigning a protocol class to a variable is allowed if it is not explicitly
+typed, and such assignment creates a type alias. For non-protocol classes,
+the behavior of ``Type[]`` is not changed.
+
+
+``NewType()`` and type aliases
+------------------------------
+
+Protocols are essentially anonymous. To emphasize this point, static type
+checkers might refuse protocol classes inside ``NewType()`` to avoid an
+illusion that a distinct type is provided::
+
+  form typing import NewType , Protocol, Iterator
+
+  class Id(Protocol):
+      code: int
+      secrets: Iterator[bytes]
+
+  UserId = NewType('UserId', Id)  # Error, can't provide distinct type
+
+On the contrary, type aliases are fully supported, including generic type
+aliases::
+
+  from typing import TypeVar, Reversible, Iterable, Sized
+
+  T = TypeVar('T')
+  CompatReversible = Union[Reversible[T], All[Iterable[T], Sized]]
+
+
+.. _discussion:
+
+Narrowing types by ``isinstance()``
+-----------------------------------
+
+Protocol types will implement special instance and class checks, similar to
+how ``Iterable`` and other ABCs in ``collections.abc`` and ``typing`` already
+do it, but this is limited to non-generic and unsubscripted generic protocols
+(``Iterable`` is statically equivalent to ``Iterable[Any]`).
+The metaclass of ``typing.Protocol`` will automatically add
+a ``__subclasshook__()`` that provides the same semantics
+for class and instance checks as for ``collections.abc`` classes::
+
+  from typing import runtime, Protocol
+
+  class Closeable(Protocol):
+      @abstractmethod
+      def close(self):
+          ...
+
+  assert isinstance(open('some/file'), Closeable)
+
+Static type checkers will understand ``isinstance(x, Proto)`` and
+``issubclass(C, Proto)`` for protocols (as they already do
+for ``Iterable`` etc.). Static type checkers will narrow types after such
+checks by the type erased ``Proto`` (i.e. with all variables having type
+``Any`` and all methods having type ``Callable[..., Any]``).
+Note that ``isinstance(x, Proto[int])`` etc. will always fail in agreement
+with PEP 484. Examples::
+
+  from typing import Iterable, Iterator, Sequence
+
+  def process(items: Iterable[int]) -> None:
+      if isinstance(items, Iterator):
+          # 'items' have type 'Iterator[int]' here
+      elif isinstance(items, Sequence[int]):
+          # Error! Can't use 'isinstance()' with subscripted protocols
+
+
+Using Protocols in Python 3.5 and Earlier
```

Comment:
I think this could be used the same way for 2.7 - 3.5, modulo the fact that function type comments should be used in 2.7.